### PR TITLE
[`flake8-pyi`] Improve autofix safety for `redundant-none-literal` (PYI061)

### LINKED ISSUE #14567: Fixes for PYI061 and RUF020 can generate `None | None`
The fixes for [`redundant-none-literal` (PYI061)](https://docs.astral.sh/ruff/rules/redundant-none-literal/) and [`never-union` (RUF020)](https://docs.astral.sh/ruff/rules/never-union/) in Ruff 0.8.0 can generate `None | None`, which raises an error at runtime.

PYI061 example:
```console
$ cat pyi061.py
from typing import Literal
x: Literal[None] | None

$ ruff check --isolated --preview --select PYI061 pyi061.py --fix
Found 1 error (1 fixed, 0 remaining).

$ cat pyi061.py
from typing import Literal
x: None | None

$ python pyi061.py 
Traceback (most recent call last):
  File "pyi061.py", line 2, in <module>
    x: None | None
       ~~~~~^~~~~~
TypeError: unsupported operand type(s) for |: 'NoneType' and 'NoneType'
```

RUF020 example:
```console
$ cat ruf020.py
from typing import Never
x: None | Never | None

$ ruff check --isolated --select RUF020 ruf020.py --fix
Found 1 error (1 fixed, 0 remaining).

$ cat ruf020.py
from typing import Never
x: None | None

$ python ruf020.py
Traceback (most recent call last):
  File "ruf020.py", line 2, in <module>
    x: None | None
       ~~~~~^~~~~~
TypeError: unsupported operand type(s) for |: 'NoneType' and 'NoneType'
```

### PR DESCRIPTION
**Pull request by sbrugman**
## Summary

Partially resolves #14567

## Test Plan

Added regression tests.

### CODE REVIEW

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_none_literal.rs [hunk 1]
```diff
@@ -87,8 +90,34 @@ pub(crate) fn redundant_none_literal<'a>(checker: &mut Checker, literal_expr: &'
     let fix = if other_literal_elements_seen {
         None
     } else {
+        // Avoid producing syntax errors when `Literal[None] | None` would be fixed to
+        // `None | None`. Instead fix to `None`. No action needed from `typing.Union`,
+        // as `Union[None, None]` is valid syntax.
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_none_literal.rs:95]: `None | None` is invalid _Python_, but it's valid _syntax_ (the code compiles, it just raises an exception!)

```suggestion
        // as `Union[None, None]` is valid Python.
```

**Review by AlexWaygood — DISMISSED** [review]
Great, thanks for the quick fix!

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_none_literal.rs [hunk 2]
```diff
@@ -87,8 +90,34 @@ pub(crate) fn redundant_none_literal<'a>(checker: &mut Checker, literal_expr: &'
     let fix = if other_literal_elements_seen {
         None
     } else {
+        // Avoid producing syntax errors when `Literal[None] | None` would be fixed to
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_none_literal.rs:93]:
```suggestion
        // Avoid producing code that would raise an exception
        // when `Literal[None] | None` would be fixed to
```

File: crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_none_literal.rs [hunk 3]
```diff
@@ -87,8 +90,35 @@ pub(crate) fn redundant_none_literal<'a>(checker: &mut Checker, literal_expr: &'
     let fix = if other_literal_elements_seen {
         None
     } else {
+        // Avoid producing code that would raise an exception
+        // when `Literal[None] | None` would be fixed to
+        // `None | None`. Instead fix to `None`. No action needed from `typing.Union`,
+        // as `Union[None, None]` is valid Python.
```
Comment (AlexWaygood) [inline crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_none_literal.rs:96]:
```suggestion
        // Avoid producing code that would raise an exception when 
        // `Literal[None] | None` would be fixed to `None | None`.
        // Instead fix to `None`. No action needed for `typing.Union`,
        // as `Union[None, None]` is valid Python.
```

**Comment by sbrugman** [comment]
I'll fix the sandwiching case for RUF020 and PYI061 in a follow-up PR.

**Comment by AlexWaygood** [comment]
I just took a look at it, and the sandwiching case seems like it might be harder to fix, at least without copying over some of the logic from PYI016 (or moving it to a common module that both rules import). I'd be okay if we said we should simply not offer a fix in that situation. We could also put a note in the docs for this rule that we recommend also enabling PYI016 if you enable this rule. What do you think?

**Comment by AlexWaygood** [comment]
It might also be worth adding this as a test snippet with a big comment next to it:

```py
a: int | Literal[None] | None
```

This snippet still gets autofixed to

```py
a: int | None | None
```

But that's actually okay! That union desugars to `(int | None) | None`, and the left-hand-side operand there is a `types.UnionType` instance, which has a `__or__` operator that accepts `None`.

**Comment by sbrugman** [comment]
The catch is that not offering a fix in cases where we detect a PEP604-style `None` union sandwich is almost equally difficult as fixing it. 

Guaranteeing the fix is safe requires checking the entire top-level union. Simply checking the parent or grandparent expression is not enough (although these will never occur in reality):
```python
d: None | ((None | Literal[None]) | None) | None
```

I'd opt to acknowledge this as a "known limitation" in the docs and to add these test cases for posterity (red-knot).

Is there any precedent of marking a rule as Unsafe when another rule is disabled?

**Comment by AlexWaygood** [comment]
I think it wouldn't be _too_ complex to detect if there are any bare `None`s in the union, and then avoid adding an autofix if so. Something like this?

```diff
--- a/crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_none_literal.rs
+++ b/crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_none_literal.rs
@@ -1,7 +1,10 @@
 use ruff_diagnostics::{Applicability, Diagnostic, Edit, Fix, FixAvailability, Violation};
 use ruff_macros::{derive_message_formats, violation};
 use ruff_python_ast::{Expr, ExprBinOp, ExprNoneLiteral, Operator};
-use ruff_python_semantic::analyze::typing::traverse_literal;
+use ruff_python_semantic::{
+    analyze::typing::{traverse_literal, traverse_union},
+    SemanticModel,
+};
 use ruff_text_size::Ranged;
 
 use smallvec::SmallVec;
@@ -90,35 +93,14 @@ pub(crate) fn redundant_none_literal<'a>(checker: &mut Checker, literal_expr: &'
     let fix = if other_literal_elements_seen {
         None
     } else {
-        // Avoid producing code that would raise an exception when
-        // `Literal[None] | None` would be fixed to `None | None`.
-        // Instead fix to `None`. No action needed for `typing.Union`,
-        // as `Union[None, None]` is valid Python.
-        // See https://github.com/astral-sh/ruff/issues/14567.
-        let replacement_range = if let Some(Expr::BinOp(ExprBinOp {
-            left,
-            op: Operator::BitOr,
-            right,
-            range: parent_range,
-        })) = checker.semantic().current_expression_parent()
-        {
-            if matches!(**left, Expr::NoneLiteral(_)) || matches!(**right, Expr::NoneLiteral(_)) {
-                *parent_range
-            } else {
-                literal_expr.range()
-            }
-        } else {
-            literal_expr.range()
-        };
-
-        Some(Fix::applicable_edit(
-            Edit::range_replacement("None".to_string(), replacement_range),
-            if checker.comment_ranges().intersects(literal_expr.range()) {
+        create_fix_edit(checker.semantic(), literal_expr).map(|edit| {
+            let applicability = if checker.comment_ranges().intersects(literal_expr.range()) {
                 Applicability::Unsafe
             } else {
                 Applicability::Safe
-            },
-        ))
+            };
+            Fix::applicable_edit(edit, applicability)
+        })
     };
 
     for none_expr in none_exprs {
@@ -134,3 +116,43 @@ pub(crate) fn redundant_none_literal<'a>(checker: &mut Checker, literal_expr: &'
         checker.diagnostics.push(diagnostic);
     }
 }
+
+fn create_fix_edit(semantic: &SemanticModel, literal_expr: &Expr) -> Option<Edit> {
+    // Avoid producing code that would raise an exception when
+    // `Literal[None] | None` would be fixed to `None | None`.
+    // Instead fix to `None`. No action needed for `typing.Union`,
+    // as `Union[None, None]` is valid Python.
+    // See https://github.com/astral-sh/ruff/issues/14567.
+    let mut enclosing_union = None;
+    let mut parent_expr = semantic.current_expression_parent();
+    while let Some(Expr::BinOp(ExprBinOp {
+        op: Operator::BitOr,
+        ..
+    })) = parent_expr
+    {
+        enclosing_union = parent_expr;
+        parent_expr = semantic.current_expression_parent();
+    }
+
+    let mut is_union_with_bare_none = false;
+    if let Some(enclosing_union) = enclosing_union {
+        traverse_union(
+            &mut |expr, _| {
+                if matches!(expr, Expr::NoneLiteral(_)) {
+                    is_union_with_bare_none = true;
+                }
+            },
+            semantic,
+            enclosing_union,
+        );
+    }
+
+    if is_union_with_bare_none {
+        None
+    } else {
+        Some(Edit::range_replacement(
+            "None".to_string(),
+            literal_expr.range(),
+        ))
+    }
+}
```

**Comment by AlexWaygood** [comment]
(Sorry, the patch I posted above had a bug in it initially. I have now tested it, and fixed the bug...)

Also, it still doesn't handle cases like `Literal[None] | Literal[None]`. Though I think that could be fixed by fiddling with the closure I pass into `traverse_union()` in that patch.

**Pull request merged**